snapcraft/internal/meta/command.py

```diff
@@ -43,130 +38,218 @@
 )
 
 
-def _get_shebang_from_file(file_path: str) -> List[str]:
+def _get_shebang_from_file(file_path: pathlib.Path) -> Optional[str]:
     """Returns the shebang from file_path."""
-    if not os.path.exists(file_path):
-        raise errors.ShebangNotFoundError()
-
     with open(file_path, "rb") as exefile:
         if exefile.read(2) != b"#!":
-            raise errors.ShebangNotFoundError()
+            return None
         shebang_line = exefile.readline().strip().decode("utf-8")
 
-    # posix is set to False to respect the quoting of variables.
-    shebang_parts = shlex.split(shebang_line, posix=False)
-    # remove the leading /usr/bin/env.
-    if shebang_parts[0] == "/usr/bin/env":
-        shebang_parts = shebang_parts[1:]
-    # or if the shebang startswith /.
-    elif shebang_parts[0].startswith("/"):
-        raise errors.ShebangInRoot()
-
-    return shebang_parts
-
-
-def _find_executable(*, binary: str, prime_dir: str) -> str:
-    found_path: Optional[str] = None
-    binary_paths = (
-        os.path.join(p, binary) for p in common.get_bin_paths(root=prime_dir)
-    )
-    for binary_path in binary_paths:
-        if _executable_is_valid(binary_path):
-            found_path = binary_path
-            break
-    else:
+    return shebang_line
+
+
+class _SnapCommandResolver:
+    def __init__(self, *, command: str, prime_path: pathlib.Path) -> None:
+        self.original_command = command
+        self._prime_path = prime_path
+        self.interpreter: Optional[str] = None
+        self.command = command
+
+    def _find_executable(self, command: str) -> Optional[pathlib.Path]:
+        """Find executable in common binary path locations.
+
+        Returns:
+          Path to executable found, whether it is in searched
+          prime directory paths, or on host using shutil.which().
+        """
+        binary_paths = (
+            pathlib.Path(p, command)
+            for p in common.get_bin_paths(root=self._prime_path)
+        )
+        for binary_path in binary_paths:
```

Comment:
How about something like
```suggestion
        primed_binary_paths = [b for b in binary_paths if _executable_is_valid(b)]
        if len(primed_binary_paths) > 1:
            logger.warning(f"Multiple binaries found for {command!r}, using {primed_binary_paths[0]!r}.")
        if primed_binary_paths:
            return primed_binary_paths[0]
```
?

Reply:
seems good to log that too, done!~